// resolve-dependencies/lib/worker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkerThread = void 0;
const child_process_1 = require("child_process");
const async_1 = require("@calebboyd/async");
class WorkerThread {
    constructor({ taskConccurency = 10 } = {}) {
        this.pending = {};
        this.child = child_process_1.fork(require.resolve('./worker-runtime'));
        this.lock = new async_1.Semaphore(taskConccurency);
        this.child.on('message', (message) => {
            this.lock.release();
            const waiter = this.pending[message.id];
            delete this.pending[message.id];
            if (message.error) {
                waiter.reject(new Error(message.error));
            }
            else {
                waiter.resolve(message.result);
            }
        });
    }
    sendMessage(message) {
        return this.lock.acquire().then((id) => {
            this.child.send({ id, ...message });
            return (this.pending[id] = async_1.createDeferred()).promise;
        });
    }
    end() {
        this.child.kill();
    }
}
exports.WorkerThread = WorkerThread;
